### PR TITLE
Cmm invariants

### PR DESCRIPTION
**Pull request by lthls**
This pull request adds an optional sanity check on the Cmm intermediate representation.

The main feature is that this check does not allow two Ccatch handlers (nested or not) to bind the same continuation in the same Cmm.fundecl structure. The first commit patches Closure to generate new continuations in places where it may duplicate code, which removes the main source of duplicate continuations.

The motivation for this additional check is not only that it is easier to reason about code when continuation identifiers are unique, but also that current work on flambda requires some changes to the backend, which are simpler if we can rely on this invariant.

The checks are disabled by default and can be activated with the `-dcmm-invariants` compilation flag, but as long as #1370 is merged I wouldn't mind reversing the default or making it non-optional.

### CODE REVIEW

**Comment by gasche** [comment]
#1370 was merged

**Comment by lthls** [comment]
There's still a test failure in asmgen when the invariants checks are turned on, but nothing that I can't fix.

I'd still like to know if other developers are interested in these invariant checks, and a review of the Closure patch.

File: asmcomp/closure.ml
Comment (stedolan) [inline asmcomp/closure.ml]: Did you mean "uniqueness" here?

**Review by stedolan — COMMENTED** [review]
I like the new Cmm invariants!

There's another change in this PR: making Closure.substitute freshen static exception names. Was this needed to ensure the new invariants hold? Did Closure previously allow static exceptions to shadow each other?

**Comment by lthls** [comment]
At the time I first submitted this PR, the static exception freshening was necessary to get the invariants to pass. It turned out that not freshening was actually a bug, which was reported as #7702 and got fixed in #1553 by adding freshening, so this part will disappear with a rebase.
This patch is fairly easy to maintain and rebase, so I'm not in a rush to get it merged but can rebase it on trunk quickly if the need arises.

**Comment by gasche** [comment]
Could we try to get this merged? It looks like it needs a rebase from @lthls from any case, and a review from someone else.

**Comment by lthls** [comment]
I've rebased the PR.
Note that the invariants checked by this PR are quite restricted (only a few basic invariants related to catch handlers), so it's unlikely that they would catch a bug without something else also breaking. I myself haven't used this code in the last year (maybe even longer).
But I like the principle of having optional invariant checks, it can help with debugging (particularly, I would expect, for newcomers).
I haven't tried to activate the checks in the CI (I consider that it would be wasteful to run these checks systematically. Running them once in a while might make sense though).

**Comment by gasche** [comment]
> I haven't tried to activate the checks in the CI (I consider that it would be wasteful to run these checks systematically. Running them once in a while might make sense though).

If none of our CI checks (github or jenkins) runs this code, then it's essentially dead code and we probably should not bother having it in the distribution -- it risks bitrotting quickly and getting silently broken. There are various options to run them less often (for example run only on pushes to trunk, not in any branch). Let's try to figure it out.

**Comment by stedolan** [comment]
The implementation and invariants look good to me, so consider this "approved" when the issue of when to run it is sorted out.

Is it particularly expensive? I agree it would be bad to make CI / general compilation noticeably slower / more resource-consuming, but the checks seem simple and fast. Is the extra time noticeable?

**Comment by lthls** [comment]
The current version is cheap (according to `-dprofile`, on a single sample run of the compilation of `typecore.ml`, I get 0.005s out of a total run time of 5.6s).
So for now at least it's reasonable to have it run systematically on one of our CI runs.
If someone wants to contribute a patch for that I can include it here.
I've also pushed a commit that allows setting the flag through `OCAMLPARAM`. I'm not sure that's the best way to activate it in CI but it's probably a good idea to have it anyway.

**Comment by gasche** [comment]
I think it would make sense to enable it for the flambda check, and also for one of our non-flambda checks (given that those correspond to two different emitters of Cmm code, if I understand correctly). I would let our CI gods, @dra27 and @shindere, comment on how to include it.

**Comment by dra27** [comment]
I'd add `--enable-cmm-invariants` to configure - it's _exactly_ the same machinery as for `--enable-flambda-invariants`. Then it's just a matter of adding that to the `CONFIG_ARG` of whichever job(s) it's wanted for in .github/workflows/build.yml.

(The `OCAMLPARAM` addition is good regardless)

**Comment by lthls** [comment]
I'm not very excited at the prospect of looking at the `configure` script... I could make a quick patch to reuse the `--enable-flambda-invariants` configuration flag to setup the initial value of `Clflags.cmm_invariants` though (although that would mean that it's activated on all our CI runs, as they all set `--enable-flambda-invariants`).

One slight annoyance of using a configuration flag is that there's no way to turn off the invariants if the compiler is configured with them. I could add `-dcmm-no-invariants` too, but this feels overkill.

**Comment by dra27** [comment]
There is already `-dflambda-no-invariants` as well - is there a particular reason for `-dcmm-invariants` not to be treated in the same way?

**Comment by shindere** [comment]
David Allsopp (2021/04/19 05:41 -0700):
> I'd add `--enable-cmm-invariants` to configure - it's _exactly_ the same machinery as for `--enable-flambda-invariants`. Then it's just a matter of adding that to the `CONFIG_ARG` of whichever job(s) it's wanted for in .github/workflows/build.yml.
>
> (The `OCAMLPARAM` addition is good regardless)

We can do that, yes. And if i'm pinged I can update Inria's CI, too.

**Comment by xavierleroy** [comment]
> I'm not very excited at the prospect of looking at the `configure` script

Try harder.

> ... I could make a quick patch to reuse the `--enable-flambda-invariants` configuration flag to setup the initial value of `Clflags.cmm_invariants` though

I was hoping the new check on Cmm code is useful also in non-flambda mode.  Is this the case?

> One slight annoyance of using a configuration flag is that there's no way to turn off the invariants if the compiler is configured with them. I could add `-dcmm-no-invariants` too, but this feels overkill.

I don't think this is necessary.  I would expect invariant checking to be turned on only by core developers and during CI testing, in which case there is no need to optionally turn the checks off

**Comment by lthls** [comment]
I've followed suggestions and added a configuration flag. I hope I patched all the places that needed to be patched.
I've enabled the checks in the full-flambda and i386-static github builds.

**Comment by dra27** [comment]
It looks good to me. One final testing thought, which may just not be practical. Is it easy enough to concoct some cmm which is expected to trigger the invariant, and so could form a test in the testsuite that the invariant does something?

**Comment by lthls** [comment]
I can write a `.cmm` file that breaks the invariants and parse it using the tools for the asmgen tests. I don't know how to translate that into an actual test though.

**Comment by shindere** [comment]
Vincent Laviron (2021/04/20 04:56 -0700):
> I can write a `.cmm` file that breaks the invariants and parse it
> using the tools for the asmgen tests. I don't know how to translate
> that into an actual test though.

If you describe what should happen in terms of which commands to run
etc. I should be able to help with translating that to an ocamltest
spec.

**Comment by lthls** [comment]
> If you describe what should happen in terms of which commands to run etc. I should be able to help with translating that to an ocamltest spec.

Thanks for the offer!
There should be only one command to run, the `codegen` tool on the input (`.cmm`) file. This command is expected to fail. The error message should be stable enough to be included as reference, but the main point is only to check that it does fail.
Unlike the regular Cmm tests, we don't need to actually link the resulting object file.

The tricky part is that we'd expect the command not to fail if the invariant checks are not activated, so we need to either:
- force the checks for all cmm tests (a small patch to `codegen` that sets `Clflags.cmm_invariants` will do the trick)
- force the checks on this test only, though a new flag passed to `codegen` (I can handle the patch to `codegen`, but I'll need help finding the correct way to pass the flag with the ocamltest spec)
- make the test conditional on the configuration value for `cmm-invariants` (either skip the test if the invariants are not enabled, or check for different results).

I'm in favour of the first solution, mostly because it's the simplest one, but I'm willing to listen to suggestions.

**Comment by shindere** [comment]
I think it'd be better to have a `cmm-invaraints` test that would pass
only if the feature is activated.

Then you test block could look something like

```
(* TEST
* cmm-invariants
** codegen
codegen_expected_exit_status = "2*
*)

Plus a check to verify the output message, if that's relevant.

The `codegen` action is already there. The variable to define its
expected exit status is not yet there but not too difficult to add and
use.

Same for the `cmm-invariants` test, it shouldn't be difficult to
implement.

Let me knonw how I can help, e.g. by pushing to your branch? Or taking
your branch as a basis, publishing to my own fork so that you can
cherry-pick from there?

**Comment by lthls** [comment]
> Let me knonw how I can help, e.g. by pushing to your branch? Or taking your branch as a basis, publishing to my own fork so that you can cherry-pick from there?

Whichever you prefer. If you want to check that the Github CI passes, pushing to my branch is probably the easiest way. If you'd prefer leaving the finishing touches to me, then publishing to your own fork is fine.

Here is the test file I'd write:
```
(* TEST
* cmm-invariants
** codegen
codegen_expected_exit_status = "2"
*)

(*
This test is here to ensure that the Cmm invariant checks
correctly catch broken Cmm programs.
*)

(function "bad_continuations" (x:int)
  (* Bad arity *)
  (catch
    (exit cont 0)
   with (cont) 1)
  (* Multiple handler definition *)
  (catch
    (exit cont 0)
   with (cont y:int) y)
  (* Exit out of scope of its handler *)
  (exit cont 0)
)
```
I can push it to my branch now (and make the CI complain), or either wait for your patches or let you commit it.

Currently, in addition to the missing `cmm-invariants` and `codegen_expected_exit_status`, the `codegen` action seems to be missing a copy or link command, as the corresponding folder doesn't contain the `.cmm` file.

**Comment by shindere** [comment]
Yes, please push the file to your branch so that I don't have to think
where to put it in the testsuite
So that the file gets copied, you may try to use the
`setup-simple-build-env` action just before calling codegen. But you
don't need to push too hard on this. Just let me know when the file is
in, please, and I'll continue the developemnt from my public fork where
you can cherry-pick from.

**Comment by lthls** [comment]
I've pushed a commit with the expected test file.

**Comment by dra27** [comment]
I agree that the first suggestion - turning the invariants for all the CMM tests - is fine (in fact, it seems sensible, as well as simpler?)

**Comment by shindere** [comment]
Vincent Laviron (2021/04/20 08:08 -0700):
> I've pushed a commit with the expected test file.

Thanks! Will work on it and report back here whendone.

**Comment by shindere** [comment]
David Allsopp (2021/04/20 08:12 -0700):
> I agree that the first suggestion - turning the invariants for all the
> CMM tests - is fine (in fact, it seems sensible, as well as simpler?)

OK, ok. So let's go this way and I'll still able to help with whatever
remains.

**Comment by lthls** [comment]
I've updated the expected test. It now doesn't depend on the configuration status, so all that remains is finding how to ensure that compiling it fails.

**Comment by shindere** [comment]
Vincent Laviron (2021/04/28 02:21 -0700):
> I've updated the expected test. It now doesn't depend on the
> configuration status, so all that remains is finding how to ensure
> that compiling it fails.

See the "codegen-exit-status" branch on shindere/ocaml.

I had to rename the variable to be ocnsistent wiht the other ones, sorry
aobut that.

Can you please gie it a try and let me know if it works as you wish?

If it does not, could you please tell me how I need to run configure or
whatever I should do to reproduce?

**Comment by lthls** [comment]
> See the "codegen-exit-status" branch on shindere/ocaml.
> I had to rename the variable to be ocnsistent wiht the other ones, sorry aobut that.

No problem, I'm not particularly attached to the previous name.

> Can you please gie it a try and let me know if it works as you wish?
> If it does not, could you please tell me how I need to run configure or whatever I should do to reproduce?

When I run the test on your branch, I still get an error (Fatal error: exception Variables.No_such_variable("codegen_exit_status")).
I don't need any particular configuration to reproduce this issue, just a basic `./configure && make world.opt && make -C testsuite/ one TEST=tests/asmgen/invariants.cmm`. Do you get the same problem ? If not, do you have any idea of what may be wrong on my side ?

I also think that with your branch the test will be reported as failing even if it has the expected status, but I think I can see what I need to patch for this, so once the previous issue is resolved I can fix this myself.

**Comment by lthls** [comment]
> When I run the test on your branch, I still get an error (Fatal error: exception Variables.No_such_variable("codegen_exit_status")).
> I don't need any particular configuration to reproduce this issue, just a basic `./configure && make world.opt && make -C testsuite/ one TEST=tests/asmgen/invariants.cmm`. Do you get the same problem ? If not, do you have any idea of what may be wrong on my side ?

Note: I can reproduce the same issue by running `ocamltest/ocamltest.opt testsuite/tests/asmgen/invariants.cmm` directly.

**Comment by shindere** [comment]
Vincent Laviron (2021/04/29 08:41 -0700):
> > See the "codegen-exit-status" branch on shindere/ocaml.
> > I had to rename the variable to be ocnsistent wiht the other ones, sorry aobut that.
>
> No problem, I'm not particularly attached to the previous name.

I was apologizing because I was the one who pointed to the wrong name,
initially.

> When I run the test on your branch, I still get an error (Fatal error: exception Variables.No_such_variable("codegen_exit_status")).
> I don't need any particular configuration to reproduce this issue, just a basic `./configure && make world.opt && make -C testsuite/ one TEST=tests/asmgen/invariants.cmm`. Do you get the same problem ? If not, do you have any idea of what may be wrong on my side ?
>
> I also think that with your branch the test will be reported as
> failing even if it has the expected status, but I think I can see what
> I need to patch for this, so once the previous issue is resolved I can
> fix this myself.

I just pushed a version that works for me.

Can you give it a try, please?

**Comment by lthls** [comment]
It works, thanks !
I pushed your commit and took this opportunity to update the Changes entry. The should now be ready for merging, as soon as someone approves it.

**Comment by shindere** [comment]
Vincent Laviron (2021/04/29 09:30 -0700):
> It works, thanks !
> I pushed your commit and took this opportunity to update the Changes
> entry. The should now be ready for merging, as soon as someone
> approves it.

I'll leave the approval to somebody else, not being an expert in CMM
things.

But I was happy to help with the test and it even fixed a bug in
ocamltest!

**Review by gasche — APPROVED** [review]
Stephen wrote "10 days ago"

> The implementation and invariants look good to me, so consider this "approved" when the issue of when to run it is sorted out.

so I'm now approving on his behalf.

File: Changes
Comment (gasche) [inline Changes]: this shoud probably have more people as reviewers (I guess at least Stephen, probably David and Sébastien as well).

**Comment by lthls** [comment]
Sorry, I forgot to actually commit the Changes update... Should be fixed now

**Comment by lthls** [comment]
And I forgot that @dra27, in addition to his helpful comments, seems to have actually reviewed the configuration changes. This oversight has been fixed.

**Comment by shindere** [comment]
Merging this broke Inria's CI because the `invariants.cmm` test was run
even when the native compiler was disabled.

Commit 9280a4b456b4194d821a722d06a42c66e552659a just pushed to trunk
should fix this.

**Comment by gasche** [comment]
Thanks! We should have remembered to run `precheck`, as this PR introduces a shiny new kind of configuration-dependent test.

**Pull request merged**
